File: ai_service/abuse_protection.py
import time
from datetime import datetime, timedelta
from collections import defaultdict
from logger import security_logger

class AbuseProtection:
    def __init__(self):
        # Rate limit storage
        self.request_counts = defaultdict(dict)
        self.ai_requests = defaultdict(dict)
        self.suspicious_ips = set()
        self.blocked_ips = set()
        
        # Configuration
        self.config = {
            'api': {
                'window': 60,  # seconds
                'max_requests': 50
            },
            'ai_generation': {
                'window': 60,  # seconds
                'max_requests': 10,
                'hourly_limit': 100
            },
            'bot_detection': {
                'requests_per_second': 10,
                'auto_block_threshold': 3
            }
        }
        
        # Last cleanup time
        self.last_cleanup = time.time()
    
    def check_api_rate_limit(self, ip, path):
        """Check general API rate limit"""
        now = time.time()
        key = f"{ip}:{path}"
        
        # Check if IP is blocked
        if ip in self.blocked_ips:
            security_logger.log_suspicious_activity(
                'Blocked IP attempted access',
                ip,
                {'path': path, 'severity': 'HIGH'}
            )
            return {
                'allowed': False,
                'reason': 'IP blocked due to abuse',
                'retry_after': 3600
            }
        
        # Initialize if not exists
        if key not in self.request_counts:
            self.request_counts[key] = {
                'count': 1,
                'reset_time': now + self.config['api']['window'],
                'first_request': now
            }
            return {'allowed': True}
        
        record = self.request_counts[key]
        
        # Reset if window expired
        if now > record['reset_time']:
            record['count'] = 1
            record['reset_time'] = now + self.config['api']['window']
            record['first_request'] = now
            return {'allowed': True}
        
        # Check for bot-like behavior
        requests_per_second = record['count'] / (now - record['first_request'])
        if requests_per_second > self.config['bot_detection']['requests_per_second']:
            self.mark_suspicious(ip, 'High request rate')
            security_logger.log_suspicious_activity(
                'Potential bot detected - high request rate',
                ip,
                {
                    'path': path,
                    'requests_per_second': round(requests_per_second, 2)
                }
            )
        
        # Check if limit exceeded
        if record['count'] >= self.config['api']['max_requests']:
            security_logger.log_rate_limit(ip, path)
            return {
                'allowed': False,
                'reason': 'Rate limit exceeded',
                'retry_after': int(record['reset_time'] - now)
            }
        
        record['count'] += 1
        return {'allowed': True}
    
    def check_ai_request(self, ip):
        """Check AI-specific rate limit"""
        now = time.time()
        
        # Check if IP is blocked
        if ip in self.blocked_ips:
            return {
                'allowed': False,
                'reason': 'IP blocked due to abuse'
            }
        
        # Initialize if not exists
        if ip not in self.ai_requests:
            self.ai_requests[ip] = {
                'count': 1,
                'reset_time': now + self.config['ai_generation']['window'],
                'hourly_count': 1,
                'hourly_reset_time': now + 3600
            }
            return {'allowed': True}
        
        record = self.ai_requests[ip]
        
        # Reset minute counter
        if now > record['reset_time']:
            record['count'] = 1
            record['reset_time'] = now + self.config['ai_generation']['window']
        elif record['count'] >= self.config['ai_generation']['max_requests']:
            security_logger.log_rate_limit(ip, 'AI generation')
            return {
                'allowed': False,
                'reason': 'AI request rate limit exceeded',
                'retry_after': int(record['reset_time'] - now)
            }
        else:
            record['count'] += 1
        
        # Reset hourly counter
        if now > record['hourly_reset_time']:
            record['hourly_count'] = 1
            record['hourly_reset_time'] = now + 3600
        elif record['hourly_count'] >= self.config['ai_generation']['hourly_limit']:
            security_logger.log_suspicious_activity(
                'Hourly AI request limit exceeded',
                ip,
                {'hourly_count': record['hourly_count']}
            )
            return {
                'allowed': False,
                'reason': 'Hourly AI request limit reached',
                'retry_after': int(record['hourly_reset_time'] - now)
            }
        else:
            record['hourly_count'] += 1
        
        return {'allowed': True}
    
    def detect_bot(self, request):
        """Detect bot-like behavior"""
        user_agent = request.headers.get('User-Agent', '')
        ip = request.remote_addr
        
        # Check for missing or suspicious user agents
        if not user_agent or len(user_agent) < 10:
            self.mark_suspicious(ip, 'Missing or suspicious user agent')
            return {
                'is_bot': True,
                'reason': 'Missing or invalid user agent',
                'confidence': 'high'
            }
        
        # Check for common bot patterns
        bot_patterns = [
            'bot', 'crawler', 'spider', 'scraper',
            'curl', 'wget', 'python-requests', 'java',
            'headless', 'phantom', 'selenium'
        ]
        
        user_agent_lower = user_agent.lower()
        for pattern in bot_patterns:
            if pattern in user_agent_lower:
                security_logger.app_logger.info(
                    f"Bot detected via user agent: {ip} - {user_agent}"
                )
                return {
                    'is_bot': True,
                    'reason': 'Bot user agent detected',
                    'confidence': 'medium'
                }
        
        # Check for missing common headers
        has_accept_language = request.headers.get('Accept-Language')
        has_accept = request.headers.get('Accept')
        has_referer = request.headers.get('Referer')
        
        suspicion_score = 0
        if not has_accept_language:
            suspicion_score += 1
        if not has_accept:
            suspicion_score += 1
        if not has_referer and request.method == 'POST':
            suspicion_score += 1
        
        if suspicion_score >= 2:
            self.mark_suspicious(ip, 'Missing common headers')
            return {
                'is_bot': True,
                'reason': 'Missing common browser headers',
                'confidence': 'medium'
            }
        
        return {'is_bot': False}
    
    def mark_suspicious(self, ip, reason):
        """Mark IP as suspicious"""
        if ip not in self.suspicious_ips:
            self.suspicious_ips.add(ip)
            security_logger.log_suspicious_activity(
                'IP marked as suspicious',
                ip,
                {'reason': reason}
            )
            
            # Auto-block after multiple suspicious activities
            suspicious_count = self.get_suspicious_count(ip)
            if suspicious_count >= self.config['bot_detection']['auto_block_threshold']:
                self.block_ip(ip, 'Multiple suspicious activities')
    
    def get_suspicious_count(self, ip):
        """Get suspicious activity count for IP"""
        count = 0
        for key, record in self.request_counts.items():
            if key.startswith(ip) and record['count'] > self.config['api']['max_requests'] * 0.8:
                count += 1
        return count
    
    def block_ip(self, ip, reason):
        """Block IP address"""
        self.blocked_ips.add(ip)
        security_logger.security_logger.warning(
            f"IP blocked: {ip} - Reason: {reason}"
        )
    
    def unblock_ip(self, ip):
        """Unblock IP address"""
        self.blocked_ips.discard(ip)
        self.suspicious_ips.discard(ip)
        security_logger.app_logger.info(f"IP unblocked: {ip}")
    
    def is_blocked(self, ip):
        """Check if IP is blocked"""
        return ip in self.blocked_ips
    
    def cleanup(self):
        """Cleanup old entries"""
        now = time.time()
        
        # Only cleanup every 5 minutes
        if now - self.last_cleanup < 300:
            return
        
        # Cleanup request counts
        keys_to_delete = []
        for key, record in self.request_counts.items():
            if now > record['reset_time'] + 300:  # 5 minutes after reset
                keys_to_delete.append(key)
        
        for key in keys_to_delete:
            del self.request_counts[key]
        
        # Cleanup AI requests
        ips_to_delete = []
        for ip, record in self.ai_requests.items():
            if now > record['hourly_reset_time'] + 300:
                ips_to_delete.append(ip)
        
        for ip in ips_to_delete:
            del self.ai_requests[ip]
        
        self.last_cleanup = now
        
        security_logger.app_logger.info(
            f"Abuse protection cleanup completed - "
            f"Request counts: {len(self.request_counts)}, "
            f"AI requests: {len(self.ai_requests)}, "
            f"Suspicious IPs: {len(self.suspicious_ips)}, "
            f"Blocked IPs: {len(self.blocked_ips)}"
        )
    
    def get_stats(self):
        """Get statistics"""
        return {
            'request_counts': len(self.request_counts),
            'ai_requests': len(self.ai_requests),
            'suspicious_ips': len(self.suspicious_ips),
            'blocked_ips': len(self.blocked_ips)
        }

# Global instance
abuse_protection = AbuseProtection()

File: server/abuseProtection.js
const logger = require('./logger');

// Advanced rate limiting with multiple tiers
class AbuseProtection {
    constructor() {
        // Rate limit storage
        this.requestCounts = new Map();
        this.loginAttempts = new Map();
        this.accountCreation = new Map();
        this.aiRequests = new Map();
        this.suspiciousIPs = new Set();
        this.blockedIPs = new Set();
        
        // Configuration
        this.config = {
            // General API rate limits
            api: {
                window: 60000, // 1 minute
                maxRequests: 100,
                burstAllowance: 20
            },
            // Login attempt limits
            login: {
                window: 900000, // 15 minutes
                maxAttempts: 5,
                lockoutDuration: 1800000 // 30 minutes
            },
            // Account creation limits
            accountCreation: {
                window: 3600000, // 1 hour
                maxAccounts: 3,
                dailyLimit: 10
            },
            // AI generation limits
            aiGeneration: {
                window: 60000, // 1 minute
                maxRequests: 10,
                hourlyLimit: 100
            },
            // Bot detection thresholds
            botDetection: {
                requestsPerSecond: 10,
                suspiciousPatterns: 5,
                autoBlockThreshold: 3
            }
        };
        
        // Cleanup old entries every 5 minutes
        setInterval(() => this.cleanup(), 300000);
    }
    
    // General API rate limiter
    checkApiRateLimit(ip, path) {
        const now = Date.now();
        const key = `${ip}:${path}`;
        
        if (this.blockedIPs.has(ip)) {
            logger.logSuspicious('Blocked IP attempted access', {
                ip,
                path,
                severity: 'HIGH'
            });
            return {
                allowed: false,
                reason: 'IP blocked due to abuse',
                retryAfter: 3600
            };
        }
        
        if (!this.requestCounts.has(key)) {
            this.requestCounts.set(key, {
                count: 1,
                resetTime: now + this.config.api.window,
                firstRequest: now
            });
            return { allowed: true };
        }
        
        const record = this.requestCounts.get(key);
        
        // Reset if window expired
        if (now > record.resetTime) {
            record.count = 1;
            record.resetTime = now + this.config.api.window;
            record.firstRequest = now;
            return { allowed: true };
        }
        
        // Check for bot-like behavior (too many requests too fast)
        const requestsPerSecond = record.count / ((now - record.firstRequest) / 1000);
        if (requestsPerSecond > this.config.botDetection.requestsPerSecond) {
            this.markSuspicious(ip, 'High request rate');
            logger.logSuspicious('Potential bot detected - high request rate', {
                ip,
                path,
                requestsPerSecond: requestsPerSecond.toFixed(2)
            });
        }
        
        // Check if limit exceeded
        if (record.count >= this.config.api.maxRequests) {
            logger.logRateLimit(ip, { path, count: record.count });
            return {
                allowed: false,
                reason: 'Rate limit exceeded',
                retryAfter: Math.ceil((record.resetTime - now) / 1000)
            };
        }
        
        record.count++;
        return { allowed: true };
    }
    
    // Login attempt rate limiter
    checkLoginAttempt(identifier) {
        const now = Date.now();
        
        if (!this.loginAttempts.has(identifier)) {
            this.loginAttempts.set(identifier, {
                attempts: 1,
                resetTime: now + this.config.login.window,
                lockedUntil: null
            });
            return { allowed: true };
        }
        
        const record = this.loginAttempts.get(identifier);
        
        // Check if account is locked
        if (record.lockedUntil && now < record.lockedUntil) {
            logger.security('Login attempt on locked account', {
                identifier,
                remainingLockTime: Math.ceil((record.lockedUntil - now) / 1000)
            });
            return {
                allowed: false,
                reason: 'Account temporarily locked due to too many failed attempts',
                retryAfter: Math.ceil((record.lockedUntil - now) / 1000)
            };
        }
        
        // Reset if window expired
        if (now > record.resetTime) {
            record.attempts = 1;
            record.resetTime = now + this.config.login.window;
            record.lockedUntil = null;
            return { allowed: true };
        }
        
        // Check if limit exceeded
        if (record.attempts >= this.config.login.maxAttempts) {
            record.lockedUntil = now + this.config.login.lockoutDuration;
            logger.security('Account locked due to failed login attempts', {
                identifier,
                attempts: record.attempts,
                lockoutDuration: this.config.login.lockoutDuration / 1000
            });
            return {
                allowed: false,
                reason: 'Too many failed login attempts. Account locked.',
                retryAfter: this.config.login.lockoutDuration / 1000
            };
        }
        
        record.attempts++;
        return { allowed: true, attemptsRemaining: this.config.login.maxAttempts - record.attempts };
    }
    
    // Reset login attempts on successful login
    resetLoginAttempts(identifier) {
        this.loginAttempts.delete(identifier);
    }
    
    // Account creation rate limiter
    checkAccountCreation(ip) {
        const now = Date.now();
        
        if (this.blockedIPs.has(ip)) {
            return {
                allowed: false,
                reason: 'IP blocked due to abuse'
            };
        }
        
        if (!this.accountCreation.has(ip)) {
            this.accountCreation.set(ip, {
                count: 1,
                resetTime: now + this.config.accountCreation.window,
                dailyCount: 1,
                dailyResetTime: now + 86400000 // 24 hours
            });
            return { allowed: true };
        }
        
        const record = this.accountCreation.get(ip);
        
        // Reset hourly counter
        if (now > record.resetTime) {
            record.count = 1;
            record.resetTime = now + this.config.accountCreation.window;
        } else if (record.count >= this.config.accountCreation.maxAccounts) {
            logger.logSuspicious('Excessive account creation attempts', {
                ip,
                count: record.count
            });
            return {
                allowed: false,
                reason: 'Too many account creation attempts. Please try again later.',
                retryAfter: Math.ceil((record.resetTime - now) / 1000)
            };
        } else {
            record.count++;
        }
        
        // Reset daily counter
        if (now > record.dailyResetTime) {
            record.dailyCount = 1;
            record.dailyResetTime = now + 86400000;
        } else if (record.dailyCount >= this.config.accountCreation.dailyLimit) {
            this.markSuspicious(ip, 'Excessive daily account creation');
            logger.logSuspicious('Daily account creation limit exceeded', {
                ip,
                dailyCount: record.dailyCount
            });
            return {
                allowed: false,
                reason: 'Daily account creation limit reached',
                retryAfter: Math.ceil((record.dailyResetTime - now) / 1000)
            };
        } else {
            record.dailyCount++;
        }
        
        return { allowed: true };
    }
    
    // AI generation rate limiter
    checkAIRequest(ip) {
        const now = Date.now();
        
        if (this.blockedIPs.has(ip)) {
            return {
                allowed: false,
                reason: 'IP blocked due to abuse'
            };
        }
        
        if (!this.aiRequests.has(ip)) {
            this.aiRequests.set(ip, {
                count: 1,
                resetTime: now + this.config.aiGeneration.window,
                hourlyCount: 1,
                hourlyResetTime: now + 3600000
            });
            return { allowed: true };
        }
        
        const record = this.aiRequests.get(ip);
        
        // Reset minute counter
        if (now > record.resetTime) {
            record.count = 1;
            record.resetTime = now + this.config.aiGeneration.window;
        } else if (record.count >= this.config.aiGeneration.maxRequests) {
            logger.logRateLimit(ip, {
                endpoint: 'AI generation',
                count: record.count
            });
            return {
                allowed: false,
                reason: 'AI request rate limit exceeded',
                retryAfter: Math.ceil((record.resetTime - now) / 1000)
            };
        } else {
            record.count++;
        }
        
        // Reset hourly counter
        if (now > record.hourlyResetTime) {
            record.hourlyCount = 1;
            record.hourlyResetTime = now + 3600000;
        } else if (record.hourlyCount >= this.config.aiGeneration.hourlyLimit) {
            logger.logSuspicious('Hourly AI request limit exceeded', {
                ip,
                hourlyCount: record.hourlyCount
            });
            return {
                allowed: false,
                reason: 'Hourly AI request limit reached',
                retryAfter: Math.ceil((record.hourlyResetTime - now) / 1000)
            };
        } else {
            record.hourlyCount++;
        }
        
        return { allowed: true };
    }
    
    // Bot detection middleware
    detectBot(req) {
        const userAgent = req.get('user-agent') || '';
        const ip = req.ip;
        
        // Check for missing or suspicious user agents
        if (!userAgent || userAgent.length < 10) {
            this.markSuspicious(ip, 'Missing or suspicious user agent');
            return {
                isBot: true,
                reason: 'Missing or invalid user agent',
                confidence: 'high'
            };
        }
        
        // Check for common bot patterns
        const botPatterns = [
            /bot/i, /crawler/i, /spider/i, /scraper/i,
            /curl/i, /wget/i, /python-requests/i, /java/i,
            /headless/i, /phantom/i, /selenium/i
        ];
        
        for (const pattern of botPatterns) {
            if (pattern.test(userAgent)) {
                logger.info('Bot detected via user agent', { ip, userAgent });
                return {
                    isBot: true,
                    reason: 'Bot user agent detected',
                    confidence: 'medium'
                };
            }
        }
        
        // Check for missing common headers
        const hasAcceptLanguage = req.get('accept-language');
        const hasAccept = req.get('accept');
        const hasReferer = req.get('referer');
        
        let suspicionScore = 0;
        if (!hasAcceptLanguage) suspicionScore++;
        if (!hasAccept) suspicionScore++;
        if (!hasReferer && req.method === 'POST') suspicionScore++;
        
        if (suspicionScore >= 2) {
            this.markSuspicious(ip, 'Missing common headers');
            return {
                isBot: true,
                reason: 'Missing common browser headers',
                confidence: 'medium'
            };
        }
        
        return { isBot: false };
    }
    
    // Mark IP as suspicious
    markSuspicious(ip, reason) {
        if (!this.suspiciousIPs.has(ip)) {
            this.suspiciousIPs.add(ip);
            logger.logSuspicious('IP marked as suspicious', { ip, reason });
            
            // Auto-block after multiple suspicious activities
            const suspiciousCount = this.getSuspiciousCount(ip);
            if (suspiciousCount >= this.config.botDetection.autoBlockThreshold) {
                this.blockIP(ip, 'Multiple suspicious activities');
            }
        }
    }
    
    // Get suspicious activity count for IP
    getSuspiciousCount(ip) {
        let count = 0;
        for (const [key, record] of this.requestCounts) {
            if (key.startsWith(ip) && record.count > this.config.api.maxRequests * 0.8) {
                count++;
            }
        }
        return count;
    }
    
    // Block IP
    blockIP(ip, reason) {
        this.blockedIPs.add(ip);
        logger.security('IP blocked', { ip, reason, severity: 'HIGH' });
    }
    
    // Unblock IP (manual intervention)
    unblockIP(ip) {
        this.blockedIPs.delete(ip);
        this.suspiciousIPs.delete(ip);
        logger.info('IP unblocked', { ip });
    }
    
    // Check if IP is blocked
    isBlocked(ip) {
        return this.blockedIPs.has(ip);
    }
    
    // Cleanup old entries
    cleanup() {
        const now = Date.now();
        
        // Cleanup request counts
        for (const [key, record] of this.requestCounts) {
            if (now > record.resetTime + 300000) { // 5 minutes after reset
                this.requestCounts.delete(key);
            }
        }
        
        // Cleanup login attempts
        for (const [key, record] of this.loginAttempts) {
            if (now > record.resetTime + 300000 && (!record.lockedUntil || now > record.lockedUntil)) {
                this.loginAttempts.delete(key);
            }
        }
        
        // Cleanup account creation
        for (const [key, record] of this.accountCreation) {
            if (now > record.dailyResetTime + 300000) {
                this.accountCreation.delete(key);
            }
        }
        
        // Cleanup AI requests
        for (const [key, record] of this.aiRequests) {
            if (now > record.hourlyResetTime + 300000) {
                this.aiRequests.delete(key);
            }
        }
        
        logger.info('Abuse protection cleanup completed', {
            requestCounts: this.requestCounts.size,
            loginAttempts: this.loginAttempts.size,
            accountCreation: this.accountCreation.size,
            aiRequests: this.aiRequests.size,
            suspiciousIPs: this.suspiciousIPs.size,
            blockedIPs: this.blockedIPs.size
        });
    }
    
    // Get statistics
    getStats() {
        return {
            requestCounts: this.requestCounts.size,
            loginAttempts: this.loginAttempts.size,
            accountCreation: this.accountCreation.size,
            aiRequests: this.aiRequests.size,
            suspiciousIPs: this.suspiciousIPs.size,
            blockedIPs: this.blockedIPs.size
        };
    }
}

module.exports = new AbuseProtection();
